go/host/enclave/state.go

```diff
@@ -0,0 +1,184 @@
+	Unavailable
+	// AwaitingSecret - enclave is waiting for host to request and provide secret
+	AwaitingSecret
+	// L1Catchup - enclave is behind on L1 data, host should submit L1 blocks to catch up
+	L1Catchup
+	// L2Catchup - enclave is behind on L2 data, host should request and submit L2 batches to catch up
+	L2Catchup
+)
+
+func (es Status) String() string {
+	return [...]string{"Live", "Disconnected", "Unavailable", "AwaitingSecret", "L1Catchup", "L2Catchup"}[es]
+}
+
+// StateTracker is the state machine for the enclave
+type StateTracker struct {
+	// status is the cached status of the enclave
+	// It is a function of the properties below and recalculated when any of them change
+	status Status
+
+	// enclave states (updated when enclave returns Status and optimistically after successful actions)
+	enclaveStatusCode common.StatusCode
+	enclaveL1Head     gethcommon.Hash
+	enclaveL2Head     gethcommon.Hash
+
+	// latest seen heads of L1 and L2 chains from external sources
+	hostL1Head gethcommon.Hash
+	hostL2Head gethcommon.Hash
+
+	m      *sync.RWMutex
+	logger gethlog.Logger
+}
+
+func NewStateTracker(logger gethlog.Logger) *StateTracker {
+	return &StateTracker{status: Disconnected, m: &sync.RWMutex{}, logger: logger}
+}
+
+func (s *StateTracker) String() string {
+	return fmt.Sprintf("StateTracker: [%s] enclave(StatusCode=%d, L1Head=%s, L2Head=%s), Host(L1Head=%s, L2Head=%s)",
+		s.status, s.enclaveStatusCode, s.enclaveL1Head, s.enclaveL2Head, s.hostL1Head, s.hostL2Head)
+}
+
+func (s *StateTracker) GetStatus() Status {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	return s.status
+}
+
+func (s *StateTracker) OnProcessedBlock(enclL1Head gethcommon.Hash) {
+	s.m.Lock()
+	s.enclaveL1Head = enclL1Head
+	s.m.Unlock()
+	s.recalculateStatus()
+}
+
+func (s *StateTracker) OnReceivedBlock(l1Head gethcommon.Hash) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.hostL1Head = l1Head
+}
+
+func (s *StateTracker) OnProcessedBatch(enclL2Head gethcommon.Hash) {
+	s.m.Lock()
+	s.enclaveL2Head = enclL2Head
+	s.m.Unlock()
+	s.recalculateStatus()
+}
+
+func (s *StateTracker) OnReceivedBatch(l2Head gethcommon.Hash) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.hostL2Head = l2Head
+}
+
+func (s *StateTracker) OnSecretProvided() {
+	s.m.Lock()
+	if s.enclaveStatusCode == common.AwaitingSecret {
+		s.enclaveStatusCode = common.Running
+	}
+	s.m.Unlock()
+	s.recalculateStatus()
+}
+
+func (s *StateTracker) OnEnclaveStatus(es common.Status) {
+	s.m.Lock()
+	s.enclaveStatusCode = es.StatusCode
+	s.enclaveL1Head = es.L1Head
+	s.enclaveL2Head = es.L2Head
+	s.m.Unlock()
+
+	s.recalculateStatus()
+}
+
+// OnDisconnected is called if the enclave is unreachable/not returning a valid Status
+func (s *StateTracker) OnDisconnected() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.setStatus(Disconnected)
+}
+
+// when enclave is operational, this method will update the status based on comparison of current chain heads with enclave heads
```

Comment:
Why execute the set status in this unified method ?
What if the execute the set status was in each of the operations ? The lock is already adquired at that stage. Would that be simpler ?

Reply:
Yeah, good shout